src/EventListener/ReadListener.php

```diff
@@ -60,61 +57,33 @@ public function __construct(CollectionDataProviderInterface $collectionDataProvi
+        if (null === $filters = ($parameters['_api_filters'] ?? null)) {
             $queryString = RequestParser::getQueryString($request);
             $filters = $queryString ? RequestParser::parseRequestParams($queryString) : null;
         }
-
-        $context = null === $filters ? [] : ['filters' => $filters];
+        $normalizationContext = [];
         if ($this->serializerContextBuilder) {
-            // Builtin data providers are able to use the serialization context to automatically add join clauses
-            $context += $normalizationContext = $this->serializerContextBuilder->createFromRequest($request, true, $attributes);
-            $request->attributes->set('_api_normalization_context', $normalizationContext);
-        }
-
-        if (isset($attributes['collection_operation_name'])) {
-            $request->attributes->set('data', $this->getCollectionData($attributes, $context));
-
-            return;
-        }
-
-        $data = [];
-
-        if ($this->identifierConverter) {
-            $context[IdentifierConverterInterface::HAS_IDENTIFIER_CONVERTER] = true;
+            $normalizationContext = $this->serializerContextBuilder->createFromRequest($request, true, $attributes);
         }
 
         try {
-            $identifiers = $this->extractIdentifiers($request->attributes->all(), $attributes);
-
-            if (isset($attributes['item_operation_name'])) {
-                $data = $this->getItemData($identifiers, $attributes, $context);
-            } elseif (isset($attributes['subresource_operation_name'])) {
-                // Legacy
-                if (null === $this->subresourceDataProvider) {
-                    throw new RuntimeException('No subresource data provider.');
-                }
-
-                $data = $this->getSubresourceData($identifiers, $attributes, $context);
-            }
-        } catch (InvalidIdentifierException $e) {
-            throw new NotFoundHttpException('Not found, because of an invalid identifier configuration', $e);
+            $data = $this->readStage->apply($attributes, $parameters, $filters, $request->getMethod(), $normalizationContext);
+        } catch (NotFoundException $e) {
+            throw new NotFoundHttpException($e->getMessage());
         }
 
         if (null === $data) {
-            throw new NotFoundHttpException('Not Found');
+            return;
+        }
+
+        if ($normalizationContext) {
+            // Builtin data providers are able to use the serialization context to automatically add join clauses
+            $request->attributes->set('_api_normalization_context', $normalizationContext);
         }
 
         $request->attributes->set('data', $data);
-        $request->attributes->set('previous_data', \is_object($data) ? clone $data : $data);
+        $request->attributes->set('previous_data', \is_object($data) && (new \ReflectionClass(\get_class($data)))->isCloneable() ? clone $data : $data);
```

Comment:
This should be backported on 2.4 as a bugfix.

Reply:
This is not really a bugfix. It's because now this method is *always* called whereas before it was not called if the operation was a collection.

Reply:
An issue has been reported for non-cloneable objects IIRC

Reply:
In this case, without this change, the tests wouldn't pass.